core/src/main/java/io/seata/core/rpc/netty/AbstractNettyRemotingClient.java

```diff
@@ -138,7 +138,21 @@ public AbstractNettyRemotingClient(NettyClientConfig nettyClientConfig, EventExe
 
     @Override
     public Object sendSyncRequest(Object msg) throws TimeoutException {
-        String serverAddress = loadBalance(getTransactionServiceGroup(), msg);
+        String xid = getXid(msg);
+        String serverAddress = null;
+        if (!NettyClientConfig.isEnableMsgServerAffinityByXid() || xid == null || xid.length() == 0) {
+            serverAddress = loadBalance(getTransactionServiceGroup(), msg);
+        } else
+        {
+            // enabled msg server affinity by xid, send all the same xid message to the same TC server
+            // TODO: When one TC server died, the recover message may come from different TC server,
+            //      such as use a new format xid such as "SEATA_RECOVER:{RECOVER_TC_URL}:OLD_XID_URL"
+            serverAddress = xid.substring(0, xid.lastIndexOf(":"));
```

Comment:
If the store mode if raft was merged. It will cause some problem.

Reply:
Ok, will the raft mode release in version 1.5? And will the raft mode carrying the high performance?
I will try to learn more about raft mode.

Reply:
Maybe 1.5 or 1.6. Here is the pr https://github.com/seata/seata/pull/3086